[JVM_IR] Fix range of locals in connection with finally blocks.

madsager opened the pull request:
For code such as:

```
try {
  var y = "y"
  for (i in 0 until 1) {
    return y
  }
} finally {
  println("finally")
}
```

The local variables `y` and `i` ended up covering the finally block as
well in the debugger.

This change splits the range of the locals so that they do
not cover the finally block.

udalov commented:
Unfortunately it fails during bootstrap when compiling this function: https://github.com/JetBrains/kotlin/blob/b27a109a1e5d0b38c374d8d34c43c98687f60755/compiler/ir/ir.interpreter/src/org/jetbrains/kotlin/ir/interpreter/IrInterpreter.kt#L710

Here's a minimized test:
```
fun box(): String {
    try {
        for (c in emptyArray<String>()) {
            g { return "Fail" }
        }
    } finally {}
    return "OK"
}

inline fun g(block: () -> Unit): Unit = block()

fun main() {
    println(box())
}
```
->
```
java.lang.IllegalStateException: Label wasn't found during iterating through instructions
	at org.jetbrains.kotlin.codegen.optimization.LabelNormalizationMethodTransformer$TransformerForMethod.getNew(LabelNormalizationMethodTransformer.kt:149)
	at org.jetbrains.kotlin.codegen.optimization.LabelNormalizationMethodTransformer$TransformerForMethod.rewriteLocalVars(LabelNormalizationMethodTransformer.kt:115)
	at org.jetbrains.kotlin.codegen.optimization.LabelNormalizationMethodTransformer$TransformerForMethod.transform(LabelNormalizationMethodTransformer.kt:39)
	at org.jetbrains.kotlin.codegen.optimization.LabelNormalizationMethodTransformer.transform(LabelNormalizationMethodTransformer.kt:28)
	at org.jetbrains.kotlin.codegen.optimization.transformer.CompositeMethodTransformer.transform(CompositeMethodTransformer.kt:25)
	at org.jetbrains.kotlin.codegen.optimization.transformer.CompositeMethodTransformer.transform(CompositeMethodTransformer.kt:25)
	at org.jetbrains.kotlin.codegen.optimization.OptimizationMethodVisitor.performTransformations(OptimizationMethodVisitor.kt:62)
	at org.jetbrains.kotlin.codegen.TransformationMethodVisitor.visitEnd(TransformationMethodVisitor.kt:70)
	... 41 more
```

madsager commented:
Thanks Alex!

I reworked the change to deal with this case as well. Since this test involved inlining, I fixed the splitting of locals in the inliner as well when generating finally blocks for non-local returns.

That, however, did not make the change simpler. Maybe @max-kammerer and/or @pyos could have a look at the inliner changes?

udalov commented:
Thanks. I'll reassign to @max-kammerer since he's more knowledgeable here, however he's on vacation until the end of the next week, so this will probably merged not earlier than to 1.5.20.

madsager commented:
Looks like this is now failing a number of non-local return test cases in `CompileKotlinAgainstInlineKotlin` tests. I'll dig into it more. I'll ping the change again when ready for review.

madsager commented:
OK, updated and teamcity prr/ build is done and looks good. @max-kammerer, this is ready for review when you have the time.

madsager commented:
Friendly ping. :)

max-kammerer commented:
@madsager Could you also add tests with non-local return from `catch` block (in inline and non-inline cases)?

max-kammerer commented:
and with non-local return from inner finaly in presense of external one

madsager commented on compiler/testData/debug/localVariables/tryFinally10.kt:
```diff
@@ -0,0 +1,55 @@
+// test.kt:19 compute: y:int=42:int
+// test.kt:20 compute: y:int=42:int, i:int=0:int
+// test.kt:4 compute: y:int=42:int, i:int=0:int, $i$f$f:int=0:int
+// test.kt:5 compute: y:int=42:int, i:int=0:int, $i$f$f:int=0:int
```
Not for this change, but for all of these inline functions we actually should remove the locals from the caller. They are not lexically visible in the inline function and therefore should not show up there when debugging. However, I would like to look at that separately.

madsager replied:
Hmm, looks like the debugger filters them out, so apparently they are not a user visible issue (and for the VM the local is there so all is well).

madsager replied:
Still might be better to do correct ranges so that the behavior does not depend on the debugger used. On the other hand if you use a debugger that does not understand the inlining locals you are probably going to have a bit of a bad time debugging this anyway. :-\

max-kammerer replied:
@madsager I can't catch your intention. `$i$f$functionName`, `$i$a$functionName`  variables are generated specially for debugger to distinguish inlined function code from inline lambda one. If any of call site finally blocks inserted into inline function than yes ranges for such variable should be corrected

madsager replied:
Oh, I wasn't talking about the inlining locals. Those are removed for the finally block and that is working as intended.

I was talking about the fact that `y` and `i` loclas are covering the instructions for the inlined code for f here. So, the locals from the method into which the inline function is inlined are in the locals table when you are in the inline function. That is not right as they are not in scope. It seems like the debugger is capable of filtering them out even though they are in the locals table. When in `f` in the intellij debugger I do not see `y` and `i`. However, in jdb they are visible for all the lines of `f`.

madsager replied:
Maybe they are there so that when in the inline function you could have two stack frames. One for caller and one for inline function callee. In the caller you would show `y` and `i`? The source maps should allow you to show full stack traces, but it doesn't look like IntelliJ is currently doing that. However, to support it, these locals would have to be there. So maybe this is working as intended?

max-kammerer replied:
@madsager It's a good question, @yanex could you answer?

yanex replied:
The current belief is that the debugger should show call-site variables in inline lambdas (where their evaluation makes sense) and shouldn't show them in inline functions (just like in a non-inline function, you don't see variables from a caller). The logic is quite straightforward there: if the debugger recognizes we're inside an inline function but not in an inlined lambda (`$i$f` appears later than `$i$a`, or `$i$a` is absent), we only show local variables with the most `$iv` at the end.

We don't show call-site variables for non-inlined lambdas. However, the Java debugger in IntelliJ does that for simple cases (for instance, when the lambda runs in the same thread), so we'll probably also implement similar behavior rather sooner than later.

Speaking about stack traces for inline functions, we don't show them now, but we've already started experimenting in this direction. Without a doubt, we'll need access to call-site variables from any place in an inline function. Yet, the current scheme is also not ideal in less trivial cases. For instance, here we lose access to the outer `a`:

```
fun foo() {
    val a = 1
    "foo".apply {
        print(a)
        val a = 2 // a = 1 is not available anymore
        print(a)
    }
}
```

madsager commented on compiler/testData/debug/localVariables/tryFinally9.kt:
```diff
@@ -0,0 +1,54 @@
+        }
+    } finally {
+        val s2 = "OK"
+        x = s2  // TODO: Why is `s2` not visible here?
```
This is a little puzzling. I would like to have a look at that separately from this change.

madsager reviewed (commented):
Thank you @max-kammerer. Adding those tests was a good call. Apparently, there is poor test coverage of nested try-finally blocks. So these tests exposed that I was no dealing with overlapping gaps in ranges (finally code inlined in other finally code) and ended up generating an invalid locals table. I have fixed that in this change by consolidating the gaps before emitting the locals. However, I do not like the offset computations I'm doing. Let me know if you think of a better way of doing it!

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt:
```diff
@@ -381,10 +390,55 @@ class ExpressionCodegen(
         get() = origin != IrDeclarationOrigin.IR_TEMPORARY_VARIABLE &&
                 origin != IrDeclarationOrigin.FOR_LOOP_ITERATOR
 
+    private fun AbstractInsnNode.computeOffset(): Int {
```
> Let me know if you think of a better way of doing it!
@madsager  I will look, looks like it's also worth to mention @pyos (he always have some fresh ideas)

madsager replied:
Yeah, we were just talking and since this should be always nested gaps inside of other gaps, maybe we can avoid generating the nested gap in the first place. However, the trivial change to do that doesn't work in the "return from nested finally with a non-local return case" (tryFinally9 test case). I'll look more next week. :)

madsager replied:
I was just holding it wrong. Updated to not generate nested gaps in the first place which resolves this issue.

max-kammerer replied:
> Apparently, there is poor test coverage of nested try-finally blocks

Yes, sadly, some parts have bad coverage. 

Thank you for adding new tests with inline in case of non-local return in catch/finally (as was mentioned in my original asking above). 
Could you also add similar test without inline functions for similar cases in catch/finally? (Sorry for misleading in original answer, it was updated in short time after posting)

madsager replied:
Thanks @max-kammerer. I added some more tests in the third commit. Are these what you were looking for?

These tests show that exception variables are not properly tracked and therefore they bleed into the finally blocks. Also, there are some interesting line stepping going on. I have added comments in the files. I would like to not fix these in this change. Once we are happy with this change, I'll file YouTrack issues for the TODOs in the test files and add them to the list of things to work on.

max-kammerer replied:
@madsager Thank you!

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt:
```diff
@@ -393,6 +423,24 @@ class ExpressionCodegen(
         }
     }
 
+    private fun addLocalVariableGapsForFinallyBlocks(
```
Not sure about naming, maybe "splitLocalVariableRangesByFinnalyBlocks"?

max-kammerer commented:
@madsager Thank you! Merged with auto-push
